test/api/test_module_construct.cpp

```diff
@@ -0,0 +1,81 @@
+#include <numeric>
+#include <migraphx/migraphx.h>
+#include <migraphx/migraphx.hpp>
+#include "test.hpp"
+
+TEST_CASE(add_op)
+{
+    migraphx::program p;
+    migraphx::module m = p.get_main_module();
+    migraphx::shape param_shape{migraphx_shape_float_type, {3, 3}};
+    auto x      = m.add_parameter("x", param_shape);
+    auto y      = m.add_parameter("y", param_shape);
+    auto add_op = migraphx::operation("add");
+    auto r      = m.add_instruction(add_op, {x, y});
+    m.add_return({r});
+    // run on ref target
+    p.compile(migraphx::target("ref"));
+    migraphx::program_parameters pp;
+    std::vector<float> x_data(9, 1);
+    std::vector<float> y_data(9, -1);
+    pp.add("x", migraphx::argument(param_shape, x_data.data()));
+    pp.add("y", migraphx::argument(param_shape, y_data.data()));
+    auto outputs = p.eval(pp);
+    auto output  = outputs[0];
+    std::vector<float> expected(9, 0);
+    CHECK(bool(output == migraphx::argument(param_shape, expected.data())));
+}
+
+TEST_CASE(if_then_else_op)
+{
+    migraphx::shape param_shape{migraphx_shape_float_type, {3, 3}};
+    migraphx::shape cond_s{migraphx_shape_bool_type};
+    auto create_program = [&]() {
+        migraphx::program p;
+        auto mm         = p.get_main_module();
+        auto cond       = mm.add_parameter("cond", cond_s);
+        auto x          = mm.add_parameter("x", param_shape);
+        auto y          = mm.add_parameter("y", param_shape);
+        auto then_mod   = p.create_module("If_0_if");
+        auto x_identity = then_mod.add_instruction(migraphx::operation("identity"), {x});
+        then_mod.add_return({x_identity});
+
+        auto else_mod   = p.create_module("If_0_else");
+        auto y_identity = else_mod.add_instruction(migraphx::operation("identity"), {y});
+        else_mod.add_return({y_identity});
+
+        auto if_ins = mm.add_instruction(migraphx::operation("if"), {cond}, {then_mod, else_mod});
+        auto get_tuple_op = migraphx::operation("get_tuple_elem", "{index: 0}");
+        auto ret          = mm.add_instruction(get_tuple_op, {if_ins});
+        mm.add_return({ret});
+        return p;
+    };
+
+    std::vector<float> x_data(9, 1);
+    std::vector<float> y_data(9, -1);
+    auto run_prog = [&](bool cond) {
+        auto p = create_program();
+        p.compile(migraphx::target("ref"));
+        migraphx::program_parameters pp;
+        char ccond        = cond;
+        auto param_shapes = p.get_parameter_shapes();
+        pp.add("cond", migraphx::argument(cond_s, &ccond));
+        pp.add("x", migraphx::argument(param_shape, x_data.data()));
+        pp.add("y", migraphx::argument(param_shape, y_data.data()));
+        auto outputs    = p.eval(pp);
+        auto output     = outputs[0];
+        float* data_ptr = reinterpret_cast<float*>(output.data());
+        std::vector<float> ret(data_ptr, data_ptr + 9);
+        return ret;
```

Comment:
This should return `outputs[0]`. No need to return a vector.

Reply:
yes, but  for some reason, when `outputs[0]` argument is returned from the lambda function, its type changes from `float_type` to `bool_type`.  Underlying data remains the same.  Since type is mismatched, i am not able to directly use `==` on `migraphx::argument`.

Reply:
That looks like a bug somewhere. Whats happens when running with ASAN? Also what does program look like after compiling for "ref"?

Reply:
```module: "main"
y = @param:y -> float_type, {3, 3}, {3, 1}
x = @param:x -> float_type, {3, 3}, {3, 1}
cond = @param:cond -> bool_type, {1}, {0}
main:@3 = if(cond), [If_0_if, If_0_else] -> [float_type, {3, 3}, {3, 1}]
main:@4 = ref::get_tuple_elem[index=0](main:@3) -> float_type, {3, 3}, {3, 1}
main:@5 = @return(main:@4)

module: "If_0_if"
If_0_if:@0 = ref::identity(x) -> float_type, {3, 3}, {3, 1}
If_0_if:@1 = @return(If_0_if:@0)

module: "If_0_else"
If_0_else:@0 = ref::identity(y) -> float_type, {3, 3}, {3, 1}
If_0_else:@1 = @return(If_0_else:@0)

This is after "ref" compilation. I ran ASAN but it seems to find unrelated memory leaks ( in libamdhip64).

Reply:
Made required changes by returning `outputs` from lambda.